File: java/src/test/java/com/grid9/core/UniformPrecisionCoordinateCompressorTest.java
package com.grid9.core;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import static org.junit.jupiter.api.Assertions.*;

import java.util.ArrayList;
import java.util.List;

/**
 * Comprehensive test suite for UniformPrecisionCoordinateCompressor
 * Validates uniform 3-meter precision across all latitudes and longitudes
 */
public class UniformPrecisionCoordinateCompressorTest {
    
    private static final double PRECISION_TOLERANCE_M = 0.1; // 10cm tolerance for precision validation
    private static final double MAX_EXPECTED_ERROR_M = 3.5; // Realistic maximum with 45-bit constraints
    private static final double TARGET_PRECISION_M = 3.0;

    @Test
    public void encode_ProducesNineCharacterString() {
        // Test various global coordinates
        double[][] testCoordinates = {
            {40.7128, -74.0060}, // New York
            {51.5074, -0.1278},  // London  
            {35.6762, 139.6503}, // Tokyo
            {-33.8688, 151.2083}, // Sydney
            {55.7558, 37.6176},  // Moscow
            {-22.9068, -43.1729}, // Rio de Janeiro
            {1.3521, 103.8198},  // Singapore
            {78.2232, 15.6267},  // Svalbard (high latitude)
            {-54.8019, -68.3030}, // Ushuaia (southern)
        };

        for (double[] coord : testCoordinates) {
            double lat = coord[0];
            double lon = coord[1];
            String encoded = UniformPrecisionCoordinateCompressor.encode(lat, lon);
            
            assertNotNull(encoded);
            assertEquals(9, encoded.length());
            
            // Verify all characters are valid base32
            for (char c : encoded.toCharArray()) {
                assertTrue("0123456789ABCDEFGHJKMNPQRSTVWXYZ".indexOf(c) >= 0,
                    "Invalid character: " + c);
            }
        }
    }

    @Test
    public void encodeWithHumanReadable_ProducesFormattedString() {
        String compact = UniformPrecisionCoordinateCompressor.encode(40.7128, -74.0060, false);
        String readable = UniformPrecisionCoordinateCompressor.encode(40.7128, -74.0060, true);
        
        assertEquals(9, compact.length());
        assertEquals(11, readable.length());
        assertEquals('-', readable.charAt(3));
        assertEquals('-', readable.charAt(7));
        
        // Both should decode to same coordinates
        double[] coords1 = UniformPrecisionCoordinateCompressor.decode(compact);
        double[] coords2 = UniformPrecisionCoordinateCompressor.decode(readable);
        
        assertEquals(coords1[0], coords2[0], 0.000001);
        assertEquals(coords1[1], coords2[1], 0.000001);
    }

    @Test
    public void decode_RoundTripAccuracy() {
        double[][] testCoordinates = {
            {40.7128, -74.0060}, // New York
            {51.5074, -0.1278},  // London  
            {35.6762, 139.6503}, // Tokyo
            {-33.8688, 151.2083}, // Sydney
            {0.0, 0.0},          // Origin
            {89.9, 0.0},         // Near North Pole
            {-89.9, 0.0},        // Near South Pole
            {0.0, 179.9},        // Near Date Line
            {0.0, -179.9},       // Near Date Line (other side)
        };

        for (double[] coord : testCoordinates) {
            double originalLat = coord[0];
            double originalLon = coord[1];
            
            String encoded = UniformPrecisionCoordinateCompressor.encode(originalLat, originalLon);
            double[] decoded = UniformPrecisionCoordinateCompressor.decode(encoded);
            double decodedLat = decoded[0];
            double decodedLon = decoded[1];
            
            // Calculate error in meters
            double distance = calculateHaversineDistance(originalLat, originalLon, decodedLat, decodedLon);
            
            assertTrue(distance <= MAX_EXPECTED_ERROR_M, 
                String.format("Round-trip error %.1fm exceeds maximum %.1fm at (%.4f, %.4f)", 
                    distance, MAX_EXPECTED_ERROR_M, originalLat, originalLon));
        }
    }

    @ParameterizedTest
    @CsvSource({
        "90, 0",    // North Pole
        "-90, 0",   // South Pole  
        "0, 180",   // Date Line
        "0, -180",  // Date Line
        "0, 0"      // Origin
    })
    public void encode_HandlesEdgeCases(double lat, double lon) {
        String encoded = UniformPrecisionCoordinateCompressor.encode(lat, lon);
        assertEquals(9, encoded.length());
        
        double[] decoded = UniformPrecisionCoordinateCompressor.decode(encoded);
        double distance = calculateHaversineDistance(lat, lon, decoded[0], decoded[1]);
        
        assertTrue(distance <= MAX_EXPECTED_ERROR_M);
    }

    @Test
    public void getActualPrecision_ReturnsUniformPrecision() {
        // Test precision at various latitudes - should be uniform with equal-area projection
        double[] testLatitudes = {-80, -60, -40, -20, 0, 20, 40, 60, 80};
        double[] testLongitudes = {-150, -90, -30, 0, 30, 90, 150};

        List<Double> precisionResults = new ArrayList<>();

        for (double lat : testLatitudes) {
            for (double lon : testLongitudes) {
                double[] precision = UniformPrecisionCoordinateCompressor.getActualPrecision(lat, lon);
                double totalError = precision[2];
                precisionResults.add(totalError);
                
                // All precision values should be similar (uniform)
                assertTrue(totalError <= MAX_EXPECTED_ERROR_M, 
                    String.format("Precision %.1fm exceeds maximum at (%.1f, %.1f)", totalError, lat, lon));
                assertTrue(totalError >= 2.0, 
                    String.format("Precision %.1fm suspiciously low at (%.1f, %.1f)", totalError, lat, lon));
            }
        }

        // Verify uniformity - calculate coefficient of variation
        double meanPrecision = precisionResults.stream().mapToDouble(Double::doubleValue).average().orElse(0.0);
        
        double variance = precisionResults.stream()
            .mapToDouble(p -> Math.pow(p - meanPrecision, 2))
            .average().orElse(0.0);
        
        double stdDev = Math.sqrt(variance);
        double coefficientOfVariation = stdDev / meanPrecision;

        // Coefficient of variation should be reasonable (longitude varies with latitude)
        assertTrue(coefficientOfVariation < 0.3, 
            String.format("Precision variation coefficient %.3f indicates non-uniform precision", coefficientOfVariation));
    }

    @Test
    public void precisionComparison_BetterThanOldAlgorithm() {
        // Compare precision at high latitudes where old algorithm was poor
        double[][] highLatitudeLocations = {
            {70.0, 0.0},     // Northern Canada
            {75.0, 120.0},   // Northern Siberia  
            {65.0, -150.0},  // Northern Alaska
            {-65.0, 45.0},   // Antarctica
        };

        for (double[] coord : highLatitudeLocations) {
            double lat = coord[0];
            double lon = coord[1];
            
            double[] precision = UniformPrecisionCoordinateCompressor.getActualPrecision(lat, lon);
            double totalError = precision[2];
            
            // New algorithm should maintain good precision even at high latitudes
            assertTrue(totalError <= 5.0, 
                String.format("High latitude precision %.1fm is poor at (%.1f, %.1f)", totalError, lat, lon));
                
            // Test actual encoding accuracy
            String encoded = UniformPrecisionCoordinateCompressor.encode(lat, lon);
            double[] decoded = UniformPrecisionCoordinateCompressor.decode(encoded);
            double actualError = calculateHaversineDistance(lat, lon, decoded[0], decoded[1]);
            
            assertTrue(actualError <= MAX_EXPECTED_ERROR_M,
                String.format("High latitude actual error %.1fm exceeds maximum at (%.1f, %.1f)", actualError, lat, lon));
        }
    }

    @Test
    public void calculateDistance_AccurateResults() {
        // Test known distances
        String nyc = UniformPrecisionCoordinateCompressor.encode(40.7128, -74.0060);
        String london = UniformPrecisionCoordinateCompressor.encode(51.5074, -0.1278);
        
        double calculatedDistance = UniformPrecisionCoordinateCompressor.calculateDistance(nyc, london);
        double expectedDistance = 5570224; // Actual NYC to London distance in meters
        
        // Allow 1% tolerance due to encoding precision
        double tolerance = expectedDistance * 0.01;
        assertTrue(Math.abs(calculatedDistance - expectedDistance) <= tolerance,
            String.format("Distance calculation error: expected %.0fm, got %.0fm", 
                expectedDistance, calculatedDistance));
    }

    @Test
    public void isValidEncoding_ValidatesCorrectly() {
        // Valid codes
        assertTrue(UniformPrecisionCoordinateCompressor.isValidEncoding("Q7KH2BBYF"));
        assertTrue(UniformPrecisionCoordinateCompressor.isValidEncoding("Q7K-H2B-BYF"));
        assertTrue(UniformPrecisionCoordinateCompressor.isValidEncoding("000000000"));
        
        // Invalid codes
        assertFalse(UniformPrecisionCoordinateCompressor.isValidEncoding(null));
        assertFalse(UniformPrecisionCoordinateCompressor.isValidEncoding(""));
        assertFalse(UniformPrecisionCoordinateCompressor.isValidEncoding("Q7KH2BBY")); // Too short
        assertFalse(UniformPrecisionCoordinateCompressor.isValidEncoding("Q7KH2BBYFF")); // Too long
        assertFalse(UniformPrecisionCoordinateCompressor.isValidEncoding("Q7KH2BBYF!")); // Invalid char
        assertFalse(UniformPrecisionCoordinateCompressor.isValidEncoding("Q7KH2BBYFI")); // Invalid char (I)
        assertFalse(UniformPrecisionCoordinateCompressor.isValidEncoding("Q7K-H2B-BY")); // Wrong format
    }

    @Test
    public void formatForHumans_FormatsCorrectly() {
        String code = "Q7KH2BBYF";
        String formatted = UniformPrecisionCoordinateCompressor.formatForHumans(code);
        assertEquals("Q7K-H2B-BYF", formatted);
        
        // Test exceptions
        assertThrows(IllegalArgumentException.class, () ->
            UniformPrecisionCoordinateCompressor.formatForHumans(null));
        assertThrows(IllegalArgumentException.class, () ->
            UniformPrecisionCoordinateCompressor.formatForHumans("Q7KH2BBY"));
    }

    @Test
    public void removeFormatting_RemovesFormattingCorrectly() {
        String formatted = "Q7K-H2B-BYF";
        String unformatted = UniformPrecisionCoordinateCompressor.removeFormatting(formatted);
        assertEquals("Q7KH2BBYF", unformatted);
        
        // Test already unformatted
        assertEquals("Q7KH2BBYF", UniformPrecisionCoordinateCompressor.removeFormatting("Q7KH2BBYF"));
        
        // Test exceptions
        assertThrows(IllegalArgumentException.class, () ->
            UniformPrecisionCoordinateCompressor.removeFormatting(null));
        assertThrows(IllegalArgumentException.class, () ->
            UniformPrecisionCoordinateCompressor.removeFormatting("Q7K-H2B-BY"));
    }

    @Test
    public void isFormattedForHumans_DetectsFormattingCorrectly() {
        assertTrue(UniformPrecisionCoordinateCompressor.isFormattedForHumans("Q7K-H2B-BYF"));
        assertFalse(UniformPrecisionCoordinateCompressor.isFormattedForHumans("Q7KH2BBYF"));
        assertFalse(UniformPrecisionCoordinateCompressor.isFormattedForHumans(null));
        assertFalse(UniformPrecisionCoordinateCompressor.isFormattedForHumans("Q7K-H2B-BY"));
        assertFalse(UniformPrecisionCoordinateCompressor.isFormattedForHumans("Q7KH2B-BYF"));
    }

    @Test
    public void getNeighbors_ReturnsNeighboringCodes() {
        String center = UniformPrecisionCoordinateCompressor.encode(40.7128, -74.0060);
        String[] neighbors = UniformPrecisionCoordinateCompressor.getNeighbors(center);
        
        // Should return up to 8 neighbors (can be fewer at edges)
        assertTrue(neighbors.length <= 8);
        assertTrue(neighbors.length > 0);
        
        // No neighbor should be the center itself
        for (String neighbor : neighbors) {
            assertNotEquals(center, neighbor);
            assertTrue(UniformPrecisionCoordinateCompressor.isValidEncoding(neighbor));
        }
        
        // Test at pole (edge case)
        String pole = UniformPrecisionCoordinateCompressor.encode(89.0, 0.0);
        String[] poleNeighbors = UniformPrecisionCoordinateCompressor.getNeighbors(pole);
        assertTrue(poleNeighbors.length > 0);
    }

    @Test
    public void encode_ThrowsOnInvalidInput() {
        // Latitude out of range
        assertThrows(IllegalArgumentException.class, () ->
            UniformPrecisionCoordinateCompressor.encode(91.0, 0.0));
        assertThrows(IllegalArgumentException.class, () ->
            UniformPrecisionCoordinateCompressor.encode(-91.0, 0.0));
        
        // Longitude out of range
        assertThrows(IllegalArgumentException.class, () ->
            UniformPrecisionCoordinateCompressor.encode(0.0, 181.0));
        assertThrows(IllegalArgumentException.class, () ->
            UniformPrecisionCoordinateCompressor.encode(0.0, -181.0));
    }

    @Test
    public void decode_ThrowsOnInvalidInput() {
        // Null/empty input
        assertThrows(IllegalArgumentException.class, () ->
            UniformPrecisionCoordinateCompressor.decode(null));
        assertThrows(IllegalArgumentException.class, () ->
            UniformPrecisionCoordinateCompressor.decode(""));
        
        // Wrong length
        assertThrows(IllegalArgumentException.class, () ->
            UniformPrecisionCoordinateCompressor.decode("Q7KH2BBY"));
        assertThrows(IllegalArgumentException.class, () ->
            UniformPrecisionCoordinateCompressor.decode("Q7KH2BBYFF"));
        
        // Invalid characters
        assertThrows(IllegalArgumentException.class, () ->
            UniformPrecisionCoordinateCompressor.decode("Q7KH2BBYF!"));
        assertThrows(IllegalArgumentException.class, () ->
            UniformPrecisionCoordinateCompressor.decode("Q7KH2BBYFI"));
    }

    @Test
    public void caseInsensitive_DecodingWorks() {
        String upper = "Q7KH2BBYF";
        String lower = "q7kh2bbyf";
        String mixed = "Q7kH2bByF";
        
        double[] coords1 = UniformPrecisionCoordinateCompressor.decode(upper);
        double[] coords2 = UniformPrecisionCoordinateCompressor.decode(lower);
        double[] coords3 = UniformPrecisionCoordinateCompressor.decode(mixed);
        
        assertEquals(coords1[0], coords2[0], 0.000001);
        assertEquals(coords1[1], coords2[1], 0.000001);
        assertEquals(coords1[0], coords3[0], 0.000001);
        assertEquals(coords1[1], coords3[1], 0.000001);
    }

    @Test
    public void realWorldLocations_ProduceExpectedCodes() {
        // Test documented examples from README
        double[] nycCoords = {40.7128, -74.0060};
        String nycCode = UniformPrecisionCoordinateCompressor.encode(nycCoords[0], nycCoords[1]);
        assertEquals("Q7KH2BBYF", nycCode);
        
        double[] londonCoords = {51.5074, -0.1278};
        String londonCode = UniformPrecisionCoordinateCompressor.encode(londonCoords[0], londonCoords[1]);
        assertEquals("S50MBZX2Y", londonCode);
        
        double[] tokyoCoords = {35.6762, 139.6503};
        String tokyoCode = UniformPrecisionCoordinateCompressor.encode(tokyoCoords[0], tokyoCoords[1]);
        assertEquals("PAYMZ39T7", tokyoCode);
    }

    @Test
    public void performance_EncodingIsReasonablyFast() {
        // Simple performance test - should complete in reasonable time
        long startTime = System.currentTimeMillis();
        
        for (int i = 0; i < 10000; i++) {
            double lat = -90 + (180.0 * i / 10000);
            double lon = -180 + (360.0 * i / 10000);
            String encoded = UniformPrecisionCoordinateCompressor.encode(lat, lon);
            assertNotNull(encoded);
        }
        
        long elapsed = System.currentTimeMillis() - startTime;
        assertTrue(elapsed < 1000, "10,000 encodings took too long: " + elapsed + "ms");
    }

    /**
     * Calculate Haversine distance between two lat/lon points
     */
    private double calculateHaversineDistance(double lat1, double lon1, double lat2, double lon2) {
        double R = 6371000.0; // Earth radius in meters
        double dLat = (lat2 - lat1) * Math.PI / 180.0;
        double dLon = (lon2 - lon1) * Math.PI / 180.0;
        double a = Math.sin(dLat/2) * Math.sin(dLat/2) +
                  Math.cos(lat1 * Math.PI / 180.0) * Math.cos(lat2 * Math.PI / 180.0) *
                  Math.sin(dLon/2) * Math.sin(dLon/2);
        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));
        return R * c;
    }
}

File: javascript/test/UniformPrecisionCoordinateCompressor.test.js
const UniformPrecisionCoordinateCompressor = require('../src/UniformPrecisionCoordinateCompressor');

describe('UniformPrecisionCoordinateCompressor', () => {
    const PRECISION_TOLERANCE_M = 0.1; // 10cm tolerance for precision validation
    const MAX_EXPECTED_ERROR_M = 3.5; // Realistic maximum with 45-bit constraints
    const TARGET_PRECISION_M = 3.0;

    describe('encode', () => {
        test('produces nine character string', () => {
            // Test various global coordinates
            const testCoordinates = [
                { lat: 40.7128, lon: -74.0060 }, // New York
                { lat: 51.5074, lon: -0.1278 },  // London  
                { lat: 35.6762, lon: 139.6503 }, // Tokyo
                { lat: -33.8688, lon: 151.2083 }, // Sydney
                { lat: 55.7558, lon: 37.6176 },  // Moscow
                { lat: -22.9068, lon: -43.1729 }, // Rio de Janeiro
                { lat: 1.3521, lon: 103.8198 },  // Singapore
                { lat: 78.2232, lon: 15.6267 },  // Svalbard (high latitude)
                { lat: -54.8019, lon: -68.3030 }, // Ushuaia (southern)
            ];

            testCoordinates.forEach(({ lat, lon }) => {
                const encoded = UniformPrecisionCoordinateCompressor.encode(lat, lon);
                
                expect(encoded).toBeDefined();
                expect(encoded).toHaveLength(9);
                
                // Verify all characters are valid base32
                for (const c of encoded) {
                    expect('0123456789ABCDEFGHJKMNPQRSTVWXYZ').toContain(c);
                }
            });
        });

        test('produces consistent results for same coordinates', () => {
            const lat = 40.7128;
            const lon = -74.0060;
            
            const encoded1 = UniformPrecisionCoordinateCompressor.encode(lat, lon);
            const encoded2 = UniformPrecisionCoordinateCompressor.encode(lat, lon);
            
            expect(encoded1).toBe(encoded2);
        });

        test('produces different results for different coordinates', () => {
            const encoded1 = UniformPrecisionCoordinateCompressor.encode(40.7128, -74.0060); // NYC
            const encoded2 = UniformPrecisionCoordinateCompressor.encode(51.5074, -0.1278);  // London
            
            expect(encoded1).not.toBe(encoded2);
        });

        test('handles boundary coordinates', () => {
            const boundaryCoords = [
                { lat: -90, lon: -180 },  // Southwest corner
                { lat: -90, lon: 180 },   // Southeast corner  
                { lat: 90, lon: -180 },   // Northwest corner
                { lat: 90, lon: 180 },    // Northeast corner
                { lat: 0, lon: 0 },       // Origin
                { lat: 0, lon: 180 },     // Antimeridian
                { lat: 0, lon: -180 },    // Antimeridian (other side)
            ];

            boundaryCoords.forEach(({ lat, lon }) => {
                const encoded = UniformPrecisionCoordinateCompressor.encode(lat, lon);
                expect(encoded).toHaveLength(9);
                expect(UniformPrecisionCoordinateCompressor.isValidEncoding(encoded)).toBe(true);
            });
        });

        test('throws error for invalid latitude', () => {
            expect(() => {
                UniformPrecisionCoordinateCompressor.encode(91, 0);
            }).toThrow('Latitude must be between -90 and 90');

            expect(() => {
                UniformPrecisionCoordinateCompressor.encode(-91, 0);
            }).toThrow('Latitude must be between -90 and 90');
        });

        test('throws error for invalid longitude', () => {
            expect(() => {
                UniformPrecisionCoordinateCompressor.encode(0, 181);
            }).toThrow('Longitude must be between -180 and 180');

            expect(() => {
                UniformPrecisionCoordinateCompressor.encode(0, -181);
            }).toThrow('Longitude must be between -180 and 180');
        });

        test('supports human readable format', () => {
            const encoded = UniformPrecisionCoordinateCompressor.encode(40.7128, -74.0060, true);
            
            expect(encoded).toHaveLength(11);
            expect(encoded[3]).toBe('-');
            expect(encoded[7]).toBe('-');
            expect(encoded).toMatch(/^[0-9A-Z]{3}-[0-9A-Z]{3}-[0-9A-Z]{3}$/);
        });
    });

    describe('decode', () => {
        test('decodes to approximate original coordinates', () => {
            const testCoordinates = [
                { lat: 40.7128, lon: -74.0060 }, // New York
                { lat: 51.5074, lon: -0.1278 },  // London  
                { lat: 35.6762, lon: 139.6503 }, // Tokyo
                { lat: 0, lon: 0 },               // Origin
            ];

            testCoordinates.forEach(({ lat, lon }) => {
                const encoded = UniformPrecisionCoordinateCompressor.encode(lat, lon);
                const decoded = UniformPrecisionCoordinateCompressor.decode(encoded);
                
                expect(Math.abs(decoded.latitude - lat)).toBeLessThan(0.01); // Within ~1km
                expect(Math.abs(decoded.longitude - lon)).toBeLessThan(0.01);
            });
        });

        test('handles human readable format', () => {
            const original = { lat: 40.7128, lon: -74.0060 };
            const encoded = UniformPrecisionCoordinateCompressor.encode(original.lat, original.lon, true);
            const decoded = UniformPrecisionCoordinateCompressor.decode(encoded);
            
            expect(Math.abs(decoded.latitude - original.lat)).toBeLessThan(0.01);
            expect(Math.abs(decoded.longitude - original.lon)).toBeLessThan(0.01);
        });

        test('throws error for invalid encoded string', () => {
            expect(() => {
                UniformPrecisionCoordinateCompressor.decode('');
            }).toThrow('Encoded string cannot be null or empty');

            expect(() => {
                UniformPrecisionCoordinateCompressor.decode('TOOLONG12');
            }).toThrow(); // Could throw either length or character error

            expect(() => {
                UniformPrecisionCoordinateCompressor.decode('SHORT');
            }).toThrow('Encoded string must be 9 characters or 11-character formatted');

            expect(() => {
                UniformPrecisionCoordinateCompressor.decode('INVALID@#');
            }).toThrow('Invalid character');
            
            expect(() => {
                UniformPrecisionCoordinateCompressor.decode('INVALIDOL'); // Contains O and L which are excluded
            }).toThrow('Invalid character');
        });

        test('round trip consistency', () => {
            const testCoordinates = [
                { lat: 40.7128, lon: -74.0060 },
                { lat: 0, lon: 0 },
                { lat: -45.123, lon: 67.456 },
                { lat: 78.22, lon: -123.45 }
            ];

            testCoordinates.forEach(({ lat, lon }) => {
                const encoded = UniformPrecisionCoordinateCompressor.encode(lat, lon);
                const decoded = UniformPrecisionCoordinateCompressor.decode(encoded);
                const reencoded = UniformPrecisionCoordinateCompressor.encode(decoded.latitude, decoded.longitude);
                
                expect(reencoded).toBe(encoded);
            });
        });
    });

    describe('precision validation', () => {
        test('achieves target precision globally', () => {
            // Test precision at various latitudes
            const testLatitudes = [-80, -45, -30, 0, 30, 45, 60, 80];
            const testLongitudes = [-180, -90, 0, 90, 180];

            testLatitudes.forEach(lat => {
                testLongitudes.forEach(lon => {
                    if (lon === 180) lon = 179.9; // Avoid boundary edge case
                    
                    const precision = UniformPrecisionCoordinateCompressor.getActualPrecision(lat, lon);
                    
                    expect(precision.totalErrorM).toBeLessThanOrEqual(MAX_EXPECTED_ERROR_M);
                    expect(precision.xErrorM).toBeGreaterThan(0);
                    expect(precision.yErrorM).toBeGreaterThan(0);
                    expect(precision.totalErrorM).toBeGreaterThan(0);
                });
            });
        });

        test('precision improves toward poles for longitude', () => {
            const longitude = 0;
            
            const equatorPrecision = UniformPrecisionCoordinateCompressor.getActualPrecision(0, longitude);
            const polarPrecision = UniformPrecisionCoordinateCompressor.getActualPrecision(80, longitude);
            
            expect(polarPrecision.xErrorM).toBeLessThan(equatorPrecision.xErrorM);
        });
    });

    describe('distance calculation', () => {
        test('calculates correct distance between known coordinates', () => {
            const nycEncoded = UniformPrecisionCoordinateCompressor.encode(40.7128, -74.0060);
            const londonEncoded = UniformPrecisionCoordinateCompressor.encode(51.5074, -0.1278);
            
            const distance = UniformPrecisionCoordinateCompressor.calculateDistance(nycEncoded, londonEncoded);
            
            // Expected distance is approximately 5570 km
            expect(distance).toBeGreaterThan(5500000); // 5500 km
            expect(distance).toBeLessThan(5600000);    // 5600 km
        });

        test('distance from point to itself is zero', () => {
            const encoded = UniformPrecisionCoordinateCompressor.encode(40.7128, -74.0060);
            const distance = UniformPrecisionCoordinateCompressor.calculateDistance(encoded, encoded);
            
            expect(distance).toBe(0);
        });
    });

    describe('validation', () => {
        test('validates correct encodings', () => {
            const validEncodings = [
                'Q7KH2BBYF',
                'S50MBZX2Y',
                'PAYMZ39T7',
                '000000000',
                'ZZZZZZZZZ'
            ];

            validEncodings.forEach(encoded => {
                expect(UniformPrecisionCoordinateCompressor.isValidEncoding(encoded)).toBe(true);
            });
        });

        test('validates human readable format', () => {
            const validFormatted = [
                'Q7K-H2B-BYF',
                'S50-MBZ-X2Y',
                '000-000-000'
            ];

            validFormatted.forEach(encoded => {
                expect(UniformPrecisionCoordinateCompressor.isValidEncoding(encoded)).toBe(true);
            });
        });

        test('rejects invalid encodings', () => {
            const invalidEncodings = [
                '',
                'SHORT',
                'TOOLONG123',
                'INVALID@#',
                'ABC-DEF',      // Wrong format
                'ABC-DEF-GHI-J', // Too long with dashes
                null,
                undefined
            ];

            invalidEncodings.forEach(encoded => {
                expect(UniformPrecisionCoordinateCompressor.isValidEncoding(encoded)).toBe(false);
            });
        });
    });

    describe('formatting', () => {
        test('formats for human readability', () => {
            const code = 'Q7KH2BBYF';
            const formatted = UniformPrecisionCoordinateCompressor.formatForHumans(code);
            
            expect(formatted).toBe('Q7K-H2B-BYF');
            expect(formatted).toHaveLength(11);
        });

        test('removes formatting', () => {
            const formatted = 'Q7K-H2B-BYF';
            const unformatted = UniformPrecisionCoordinateCompressor.removeFormatting(formatted);
            
            expect(unformatted).toBe('Q7KH2BBYF');
            expect(unformatted).toHaveLength(9);
        });

        test('detects human readable format', () => {
            expect(UniformPrecisionCoordinateCompressor.isFormattedForHumans('Q7K-H2B-BYF')).toBe(true);
            expect(UniformPrecisionCoordinateCompressor.isFormattedForHumans('Q7KH2BBYF')).toBe(false);
            expect(UniformPrecisionCoordinateCompressor.isFormattedForHumans('ABC')).toBe(false);
        });

        test('throws error for invalid format operations', () => {
            expect(() => {
                UniformPrecisionCoordinateCompressor.formatForHumans('SHORT');
            }).toThrow('Input must be exactly 9 characters');

            expect(() => {
                UniformPrecisionCoordinateCompressor.removeFormatting('INVALID');
            }).toThrow('Input must be in XXX-XXX-XXX format');
        });
    });

    describe('neighbors', () => {
        test('generates neighboring coordinates', () => {
            const center = UniformPrecisionCoordinateCompressor.encode(40.7128, -74.0060);
            const neighbors = UniformPrecisionCoordinateCompressor.getNeighbors(center);
            
            expect(neighbors).toBeInstanceOf(Array);
            expect(neighbors.length).toBeGreaterThan(0);
            expect(neighbors.length).toBeLessThanOrEqual(8);
            
            // All neighbors should be valid encodings
            neighbors.forEach(neighbor => {
                expect(UniformPrecisionCoordinateCompressor.isValidEncoding(neighbor)).toBe(true);
                expect(neighbor).not.toBe(center); // Should not include center
            });
        });

        test('neighbors are nearby', () => {
            const center = UniformPrecisionCoordinateCompressor.encode(0, 0); // Equator
            const neighbors = UniformPrecisionCoordinateCompressor.getNeighbors(center);
            
            neighbors.forEach(neighbor => {
                const distance = UniformPrecisionCoordinateCompressor.calculateDistance(center, neighbor);
                expect(distance).toBeLessThan(10000); // Should be within 10km
                expect(distance).toBeGreaterThan(0);   // Should not be zero
            });
        });
    });

    describe('edge cases', () => {
        test('handles coordinates near poles', () => {
            const polarCoords = [
                { lat: 89.999, lon: 0 },
                { lat: -89.999, lon: 0 },
                { lat: 89, lon: 179 },
                { lat: -89, lon: -179 }
            ];

            polarCoords.forEach(({ lat, lon }) => {
                const encoded = UniformPrecisionCoordinateCompressor.encode(lat, lon);
                expect(encoded).toHaveLength(9);
                
                const decoded = UniformPrecisionCoordinateCompressor.decode(encoded);
                expect(Math.abs(decoded.latitude - lat)).toBeLessThan(1.0);
                expect(Math.abs(decoded.longitude - lon)).toBeLessThan(1.0);
            });
        });

        test('handles coordinates near antimeridian', () => {
            const antimeridianCoords = [
                { lat: 0, lon: 179.999 },
                { lat: 0, lon: -179.999 },
                { lat: 45, lon: 180 },
                { lat: -45, lon: -180 }
            ];

            antimeridianCoords.forEach(({ lat, lon }) => {
                const encoded = UniformPrecisionCoordinateCompressor.encode(lat, lon);
                expect(encoded).toHaveLength(9);
                
                const decoded = UniformPrecisionCoordinateCompressor.decode(encoded);
                expect(Math.abs(decoded.latitude - lat)).toBeLessThan(0.1);
            });
        });

        test('handles very small coordinate differences', () => {
            const base = { lat: 40.7128, lon: -74.0060 };
            const nearby = { lat: 40.7128001, lon: -74.0060001 }; // ~10cm difference
            
            const encoded1 = UniformPrecisionCoordinateCompressor.encode(base.lat, base.lon);
            const encoded2 = UniformPrecisionCoordinateCompressor.encode(nearby.lat, nearby.lon);
            
            // May or may not be the same encoding depending on grid alignment
            expect(encoded1).toHaveLength(9);
            expect(encoded2).toHaveLength(9);
        });
    });

    describe('performance characteristics', () => {
        test('encodes many coordinates efficiently', () => {
            const startTime = Date.now();
            const iterations = 1000;
            
            for (let i = 0; i < iterations; i++) {
                const lat = (Math.random() - 0.5) * 160; // -80 to 80
                const lon = (Math.random() - 0.5) * 360; // -180 to 180
                UniformPrecisionCoordinateCompressor.encode(lat, lon);
            }
            
            const endTime = Date.now();
            const duration = endTime - startTime;
            
            // Should complete 1000 operations in reasonable time (< 1 second)
            expect(duration).toBeLessThan(1000);
        });

        test('decodes many coordinates efficiently', () => {
            // Pre-generate test data
            const encodedCoords = [];
            for (let i = 0; i < 1000; i++) {
                const lat = (Math.random() - 0.5) * 160;
                const lon = (Math.random() - 0.5) * 360;
                encodedCoords.push(UniformPrecisionCoordinateCompressor.encode(lat, lon));
            }
            
            const startTime = Date.now();
            
            encodedCoords.forEach(encoded => {
                UniformPrecisionCoordinateCompressor.decode(encoded);
            });
            
            const endTime = Date.now();
            const duration = endTime - startTime;
            
            // Should complete 1000 operations in reasonable time
            expect(duration).toBeLessThan(1000);
        });
    });
});